flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkResourceManager.java

```diff
@@ -0,0 +1,647 @@
+
+				LOG.info(diagMessage);
+				LOG.info("Total number of failed containers so far: " + failedContainersSoFar);
+
+				// maxFailedContainers == -1 is infinite number of retries.
+				if (maxFailedContainers >= 0 && failedContainersSoFar > maxFailedContainers) {
+					String msg = "Stopping YARN session because the number of failed containers ("
+						+ failedContainersSoFar + ") exceeded the maximum failed containers ("
+						+ maxFailedContainers + "). This number is controlled by the '"
+						+ ConfigConstants.YARN_MAX_FAILED_CONTAINERS + "' configuration setting. "
+						+ "By default its the number of requested containers.";
+
+					LOG.error(msg);
+					self().tell(decorateMessage(new StopCluster(ApplicationStatus.FAILED, msg)),
+						ActorRef.noSender());
+
+					// no need to do anything else
+					return;
+				}
+			}
+		}
+
+		updateProgress();
+
+		// in case failed containers were among the finished containers, make
+		// sure we re-examine and request new ones
+		triggerCheckWorkers();
+	}
+
+	// ------------------------------------------------------------------------
+	//  Utilities
+	// ------------------------------------------------------------------------
+
+	private void updateProgress() {
+		final int required = getDesignatedWorkerPoolSize();
+		final int available = getNumberOfRegisteredTaskManagers() + containersInLaunch.size();
+		final float progress = (required <= 0) ? 1.0f : available / (float) required;
+
+		if (resourceManagerCallbackHandler != null) {
+			resourceManagerCallbackHandler.setCurrentProgress(progress);
+		}
+	}
+
+	/**
+	 * Converts a Flink application status enum to a YARN application status enum.
+	 * @param status The Flink application status.
+	 * @return The corresponding YARN application status.
+	 */
+	private FinalApplicationStatus getYarnStatus(ApplicationStatus status) {
+		if (status == null) {
+			return FinalApplicationStatus.UNDEFINED;
+		}
+		else {
+			switch (status) {
+				case SUCCEEDED:
+					return FinalApplicationStatus.SUCCEEDED;
+				case FAILED:
+					return FinalApplicationStatus.FAILED;
+				case CANCELED:
+					return FinalApplicationStatus.KILLED;
+				default:
+					return FinalApplicationStatus.UNDEFINED;
+			}
+		}
+	}
+
+	/**
+	 * Checks if a YARN application still has registered containers. If the application master
+	 * registered at the resource manager for the first time, this list will be empty. If the
+	 * application master registered a repeated time (after a failure and recovery), this list
+	 * will contain the containers that were previously allocated.
+	 * 
+	 * @param response The response object from the registration at the ResourceManager.
+	 * @return A list with containers from previous application attempt.
+	 */
+	private List<Container> getContainersFromPreviousAttempts(RegisterApplicationMasterResponse response) {
+		try {
+			Method m = RegisterApplicationMasterResponse.class
+				.getMethod("getContainersFromPreviousAttempts");
```

Comment:
Why did you remove the `RegisterApplicationMasterResponseReflector`? With the current implementation you will use reflection every time you call `getContainersFromPreviousAttempts`. This is inefficient and not necessary since `RegisterApplicationMasterResponse.class` won't change.